File: practice/algorithms/implementation/cavity_map/cavity_map.hs
import Control.Applicative
import Control.Monad
import System.IO

getMultipleLines :: Int -> IO [String]

getMultipleLines n
    | n <= 0 = return []
    | otherwise = do          
        x <- getLine         
        xs <- getMultipleLines (n-1)    
        let ret = (x:xs)    
        return ret

findCavityUtil grid row col n
    | row == 0 || row == n - 1 = (grid !! row) !! col
    | col == 0 || col == n - 1 = (grid !! row) !! col
    | (grid !! (row - 1)) !! col >= (grid !! row) !! col = (grid !! row) !! col
    | (grid !! (row + 1)) !! col >= (grid !! row) !! col = (grid !! row) !! col
    | (grid !! row) !! (col - 1) >= (grid !! row) !! col = (grid !! row) !! col
    | (grid !! row) !! (col + 1) >= (grid !! row) !! col = (grid !! row) !! col
    | otherwise = 'X'

findCavity grid n = [[findCavityUtil grid row col n | col <- [0..(n - 1)]] | row <- [0..(n - 1)]]

main :: IO ()
main = do
    n <- readLn :: IO Int
    grid <- getMultipleLines n
    putStr $ unlines $ findCavity grid n

File: practice/algorithms/implementation/cavity_map/cavity_map.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

char findCavity(const vector<string> & grid, int row, int col, int n)
{
    if (row == 0 || row == n - 1)
        return grid[row][col];
    if (col == 0 || col == n - 1)
        return grid[row][col];
    
    if (grid[row - 1][col] >= grid[row][col])
        return grid[row][col];
    if (grid[row + 1][col] >= grid[row][col])
        return grid[row][col];
    if (grid[row][col - 1] >= grid[row][col])
        return grid[row][col];
    if (grid[row][col + 1] >= grid[row][col])
        return grid[row][col];
    
    return 'X';
}

int main()
{
    int n;
    cin >> n;
    vector<string> grid(n);
    for(int i = 0; i < n; i++)
    {
       cin >> grid[i];
    }
    
    vector<string> rgrid(n, string(n, ' '));
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            rgrid[i][j] = findCavity(grid, i, j, n);
        }
    }
    
    for (int i = 0; i < n; i++)
    {
        cout << rgrid[i] << endl;
    }
    
    return 0;
}

